File: packages/go/http_client.go
package constellation

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"
)

const defaultTimeout = 30

// HTTPClient is a simple HTTP client for network operations
type HTTPClient struct {
	client  *http.Client
	baseURL string
}

// NewHTTPClient creates a new HTTP client
func NewHTTPClient(baseURL string, timeout int) *HTTPClient {
	if timeout <= 0 {
		timeout = defaultTimeout
	}

	return &HTTPClient{
		client: &http.Client{
			Timeout: time.Duration(timeout) * time.Second,
		},
		baseURL: strings.TrimSuffix(baseURL, "/"),
	}
}

// Get makes a GET request
func (c *HTTPClient) Get(path string, result interface{}) error {
	url := c.baseURL + path

	req, err := http.NewRequest(http.MethodGet, url, nil)
	if err != nil {
		return NewNetworkError(err.Error(), 0, "")
	}

	req.Header.Set("Accept", "application/json")

	return c.doRequest(req, result)
}

// Post makes a POST request
func (c *HTTPClient) Post(path string, body interface{}, result interface{}) error {
	url := c.baseURL + path

	jsonBody, err := json.Marshal(body)
	if err != nil {
		return NewNetworkError(fmt.Sprintf("failed to marshal body: %v", err), 0, "")
	}

	req, err := http.NewRequest(http.MethodPost, url, bytes.NewReader(jsonBody))
	if err != nil {
		return NewNetworkError(err.Error(), 0, "")
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Accept", "application/json")

	return c.doRequest(req, result)
}

func (c *HTTPClient) doRequest(req *http.Request, result interface{}) error {
	resp, err := c.client.Do(req)
	if err != nil {
		if err, ok := err.(interface{ Timeout() bool }); ok && err.Timeout() {
			return ErrRequestTimeout
		}
		return NewNetworkError(err.Error(), 0, "")
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return NewNetworkError(fmt.Sprintf("failed to read response: %v", err), resp.StatusCode, "")
	}

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return NewNetworkError(
			fmt.Sprintf("HTTP %d: %s", resp.StatusCode, http.StatusText(resp.StatusCode)),
			resp.StatusCode,
			string(body),
		)
	}

	if result != nil && len(body) > 0 {
		if err := json.Unmarshal(body, result); err != nil {
			return NewNetworkError(fmt.Sprintf("failed to unmarshal response: %v", err), 0, string(body))
		}
	}

	return nil
}

File: packages/java/main/src/main/java/io/constellationnetwork/metagraph/sdk/HttpClient.java
package io.constellationnetwork.metagraph.sdk;

import com.google.gson.Gson;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;

import java.net.URI;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;

/**
 * Simple HTTP client for network operations using Java 11 HttpClient.
 */
public class HttpClient {

    private static final int DEFAULT_TIMEOUT = 30;
    private static final Gson gson = new Gson();

    private final java.net.http.HttpClient client;
    private final String baseUrl;
    private final int timeout;

    /**
     * Create a new HTTP client.
     *
     * @param baseUrl Base URL for requests
     * @param timeout Request timeout in seconds
     */
    public HttpClient(String baseUrl, int timeout) {
        this.baseUrl = baseUrl.replaceAll("/$", "");
        this.timeout = timeout > 0 ? timeout : DEFAULT_TIMEOUT;
        this.client = java.net.http.HttpClient.newBuilder()
                .connectTimeout(Duration.ofSeconds(this.timeout))
                .build();
    }

    /**
     * Make a GET request.
     *
     * @param path Request path
     * @param responseType Type of response to deserialize
     * @return Deserialized response
     */
    public <T> T get(String path, Class<T> responseType) {
        try {
            HttpRequest request = HttpRequest.newBuilder()
                    .uri(URI.create(baseUrl + path))
                    .header("Accept", "application/json")
                    .timeout(Duration.ofSeconds(timeout))
                    .GET()
                    .build();

            return executeRequest(request, responseType);
        } catch (NetworkTypes.NetworkException e) {
            throw e;
        } catch (Exception e) {
            throw new NetworkTypes.NetworkException("Request failed: " + e.getMessage(), e);
        }
    }

    /**
     * Make a POST request.
     *
     * @param path Request path
     * @param body Request body to serialize as JSON
     * @param responseType Type of response to deserialize
     * @return Deserialized response
     */
    public <T> T post(String path, Object body, Class<T> responseType) {
        try {
            String jsonBody = gson.toJson(body);

            HttpRequest request = HttpRequest.newBuilder()
                    .uri(URI.create(baseUrl + path))
                    .header("Content-Type", "application/json")
                    .header("Accept", "application/json")
                    .timeout(Duration.ofSeconds(timeout))
                    .POST(HttpRequest.BodyPublishers.ofString(jsonBody))
                    .build();

            return executeRequest(request, responseType);
        } catch (NetworkTypes.NetworkException e) {
            throw e;
        } catch (Exception e) {
            throw new NetworkTypes.NetworkException("Request failed: " + e.getMessage(), e);
        }
    }

    private <T> T executeRequest(HttpRequest request, Class<T> responseType) throws Exception {
        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());

        int statusCode = response.statusCode();
        String body = response.body();

        if (statusCode < 200 || statusCode >= 300) {
            throw new NetworkTypes.NetworkException(
                    "HTTP " + statusCode,
                    statusCode,
                    body
            );
        }

        if (body == null || body.isEmpty()) {
            return null;
        }

        return gson.fromJson(body, responseType);
    }
}
